modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -3832,9 +3946,49 @@ def apply_func(df):  # pragma: no cover
                 result.attrs[skip_on_aligning_flag] = True
             return result
 
-        result = self._apply_func_to_range_partitioning(
-            key_columns=by,
+        if len(external_by) > 0:
+            grouper = (
+                external_by[0]
+                if len(external_by) == 1
+                else external_by[0].concat(
+                    axis=1, others=external_by[1:], how="left", sort=False
+                )
+            )
+
+            new_grouper_cols = []
+            columns_were_changed = False
+            same_columns = {}
+            # duplicated names break range-partitioning mechanism, so renaming them.
+            # original names will be reverted in the actual groupby kernel
+            for col in grouper.columns:
+                suffix = same_columns.get(col)
+                if suffix is None:
+                    same_columns[col] = 0
+                else:
+                    same_columns[col] += 1
+                    col = (
+                        (*col[:-1], f"{col[-1]}_{suffix}{duplicated_suffix}")
+                        if isinstance(col, tuple)
```

Comment:
What might be the lenght of this tuple? Why do we only add a suffix to the last element?

Reply:
> What might be the lenght of this tuple?

Tuple is a value from a MultiIndex, so it can be any length in theory. 

> Why do we only add a suffix to the last element?

It's enough and simpler to add/remove the suffix only at the last level of MultiIndex, so I did it that way